homeassistant/helpers/template.py

```diff
@@ -790,6 +826,67 @@ def random_every_time(context, values):
 class TemplateEnvironment(ImmutableSandboxedEnvironment):
     """The Home Assistant template environment."""
 
+    def __init__(self, hass):
+        """Initialise template environment."""
+        super().__init__()
+        self.hass = hass
+        self.filters['round'] = forgiving_round
+        self.filters['multiply'] = multiply
+        self.filters['log'] = logarithm
+        self.filters['sin'] = sine
+        self.filters['cos'] = cosine
+        self.filters['tan'] = tangent
+        self.filters['sqrt'] = square_root
+        self.filters['as_timestamp'] = forgiving_as_timestamp
+        self.filters['timestamp_custom'] = timestamp_custom
+        self.filters['timestamp_local'] = timestamp_local
+        self.filters['timestamp_utc'] = timestamp_utc
+        self.filters['is_defined'] = fail_when_undefined
+        self.filters['max'] = max
+        self.filters['min'] = min
+        self.filters['random'] = random_every_time
+        self.filters['base64_encode'] = base64_encode
+        self.filters['base64_decode'] = base64_decode
+        self.filters['ordinal'] = ordinal
+        self.filters['regex_match'] = regex_match
+        self.filters['regex_replace'] = regex_replace
+        self.filters['regex_search'] = regex_search
+        self.filters['regex_findall_index'] = regex_findall_index
+        self.filters['bitwise_and'] = bitwise_and
+        self.filters['bitwise_or'] = bitwise_or
+        self.globals['log'] = logarithm
+        self.globals['sin'] = sine
+        self.globals['cos'] = cosine
+        self.globals['tan'] = tangent
+        self.globals['sqrt'] = square_root
+        self.globals['pi'] = math.pi
+        self.globals['tau'] = math.pi * 2
+        self.globals['e'] = math.e
+        self.globals['float'] = forgiving_float
+        self.globals['now'] = dt_util.now
+        self.globals['utcnow'] = dt_util.utcnow
+        self.globals['as_timestamp'] = forgiving_as_timestamp
+        self.globals['relative_time'] = dt_util.get_age
+        self.globals['strptime'] = strptime
+        if hass is None:
+            return
+
+        def hassfunction(func):
+            """Wrap function that depend on hass."""
+            @wraps(func)
+            def wrapper(*args, **kwargs):
+                return func(hass, *args[1:], **kwargs)
+            return contextfunction(wrapper)
```

Comment:
Why mark it as a context function but then remove the context inside the wrapper? Could you add a comment

Reply:
Comment added. It's so that it gets freshly evaluated with each run. the context object isn't required though.